### PR TITLE
fix(provider-sdk): use send_request to handle request timeout 

### PR DESCRIPTION
**Pull request by Iceber**
## Feature or Problem

`rpc_client.request` will always use the default timeout, and even if the user sets a larger timeout, we will still be limited by it.
https://github.com/wasmCloud/wasmCloud/blob/6219096dba216c4795c4c85f91a1e01012833065/crates/provider-sdk/src/rpc_client.rs#L214-L218
https://github.com/wasmCloud/wasmCloud/blob/6219096dba216c4795c4c85f91a1e01012833065/crates/provider-sdk/src/rpc_client.rs#L251-L255

use `send_request` to send the request

### CODE REVIEW

**Comment by Iceber** [comment]
It looks like `wasmbus-rpc` needs to be fixed as well, I'll submit a new pr over there after this one is approved.

**Comment by connorsmith256** [comment]
Hey @Iceber, `request` is meant to be a low-level function used within the RPC SDK. Using a consistent timeout value is by design: providers receive this timeout value from the host as `default_rpc_timeout_ms` in the `HostData`. (This field's type is `Option<u64>`, but that is for historical reasons. The host always provides a value.)

`request` is not meant to be directly called by providers. I think it's a mistake that it is marked as public and the doc comment says it can be used for general NATS messages, because the client used by this function is configured to use an RPC-specific connection + credentials + timeout. CC @thomastaylor312). Providers should call `send` to send RPC messages.

For providers that need to send messages with a custom timeout, we have `send_timeout`, which takes a custom `timeout` argument. This should be used rarely. The only use case I know of is for the HTTP server, which allows setting a [custom timeout for handling requests](https://github.com/wasmCloud/wasmCloud/blob/main/crates/providers/http-server/src/lib.rs#L355).

I think you did find a bug, however: if a provider calls `send_timeout` with a timeout of 10s, but the RPC client's inner timeout was set to 2s, the inner future returned by `maybe_timeout` will resolve after 2s, not waiting the full timeout.

So I think the fix should be to remove the `tokio::time::timeout` wrappers from the provider SDK. The underlying NATS client is already sending requests with a timeout, so there's no way to _extend_ this timeout with tokio, which would be a more common use case than shortening the timeout.

To extend the timeout, we will need to refactor the call to `self.client.request` (note that the `maybe_timeout` around `self.client.publish` does nothing, since publishing should never take more than a few milliseconds). Instead of calling `async_nats::Client::request`, we will need to use `async_nats::Client::send_request` and pass a request with a timeout. Here is [an example](https://github.com/wasmCloud/wasmCloud/blob/main/crates/host/src/wasmbus/mod.rs#L384) where we did this in the host.

**Comment by Iceber** [comment]
> request is not meant to be directly called by providers. I think it's a mistake that it is marked as public and the doc comment says it can be used for general NATS messages, because the client used by this function is configured to use an RPC-specific connection + credentials + timeout. CC @thomastaylor312). Providers should call send to send RPC messages.

Provide a `request_timeout` function wrapper in order to ensure that `request` calls and `request_timeout` calls have consistent call logs, I will remove the **pub**.

> So I think the fix should be to remove the tokio::time::timeout wrappers from the provider SDK. The underlying NATS client is already sending requests with a timeout, so there's no way to extend this timeout with tokio, which would be a more common use case than shortening the timeout.
> To extend the timeout, we will need to refactor the call to self.client.request (note that the maybe_timeout around self.client.publish does nothing, since publishing should never take more than a few milliseconds). Instead of calling async_nats::Client::request, we will need to use async_nats::Client::send_request and pass a request with a timeout. Here is [an example](https://github.com/wasmCloud/wasmCloud/blob/main/crates/host/src/wasmbus/mod.rs#L384) where we did this in the host.

Yes, it would be more appropriate to use send_request, and I'll update this pr

File: crates/provider-sdk/src/rpc_client.rs
Comment (Iceber) [inline crates/provider-sdk/src/rpc_client.rs:260]: notes: `request` always sets the timeout, which will make the request unaffected by the timeout in the `nats client`

**Review by connorsmith256 — APPROVED** [review]
LGTM! Thanks for fixing this @Iceber. I'll wait to see if @thomastaylor312 has any other feedback before merging

EDIT: we're blocked on merging currently, as we debug an issue in CI. Stand by 🙏

**Comment by connorsmith256** [comment]
@Iceber would you mind rebasing off main? I think we've resolved the issue in CI, but it's still happening on your branch

**Review by thomastaylor312 — APPROVED** [review]
This looks all good. Merge away @connorsmith256! Thanks for the fix @Iceber

**Pull request closed without merging**
